### ISSUE TITLE
Indication of space left on device when programming micro:bit

### ISSUE THREAD

**Issue by OwenBrotherwood** [Microsoft/pxt#651]
Enhancement: Indication of space left on device when programming micro:bit
As a limited storage device, the micro:bit has a small finite resource for storage(space).
It can be envisaged that an algorithm could be created that indicates to the user that the limits are reached
For example, the mbed compiler has such an indication for a build.
<img width="625" alt="memory usage" src="https://cloud.githubusercontent.com/assets/866122/19935917/16ee1aea-a11c-11e6-905e-c8b64e806d2f.PNG">

---

I hope this is not in the code or already asked for, or my name is "mud"

**Comment by abchatra** [comment]
Note: This is not possible to implement in the web app as the browser has no access to USB devices as of today. There is no way of knowing the exact memory size of the microcontroller.

**Comment by whaleygeek** [comment]
Hi @abchatra thanks for your update. However, could I just query the memory size issue? If PXT is targetting a specific platform (micro:bit), then presumably it must know where in memory it is locating the generated code in order to build the .hex file with the memory addresses in the headers of each line of the file.

So, presumably, if you know because the platform is a micro:bit, and you know you are merging it with a pre-built .hex runtime image where the runtime image is located (you must do, in order to be able to link addresses), then presumably you can measure the range of addresses the build process is generating inside the intel .hex file, and therefore work out percentage full of flash?

I was looking at the same issue in the MicroPython builds.

Mu does a script size check at the point of building the .hex file:

https://github.com/mu-editor/mu/blob/0f05abac63319a2e15ef5cd32f846f712b07dd17/mu/logic.py#L375

However, the online Python editor has a placeholder, but it is not implemented:

https://github.com/bbcmicrobit/PythonEditor/blob/4e409467fc3643cd673c0dc12e453c1b52ff147b/python-main.js#L92

I know that it is 'easier' in Python because they have a fixed size 8K region for the script.

But it seems possible to me to know that the standard runtime is X bytes in size, the micro:bit has Y bytes, and the script generated by the build process is Z. Could it be possible therefore to pop up a message if the build process generates a .hex file that will not fit inside the micro:bit?

I appreciate that showing a gauge of free space might be much harder - but having a message pop up when it is 'full' would prevent the user downloading a large script that then does not work.

Or is this check already done in the software build process?

Thanks!

**Comment by pelikhan** [comment]
We do know the size of the runtime including the dynamic code section as we need to communicate this to the file system. We do not handle the "full" scenario yet.

Get Outlook for iOS<https://aka.ms/o0ukef>

________________________________

**Comment by OwenBrotherwood** [comment]
Any thoughts on status for this issue?

**Comment by abchatra** [comment]
Wont fix for now. Feel free to provide a pull request.

**Comment by OwenBrotherwood** [comment]
OK with Revisit label
